File: clients/ios/Lextures/Features/Reader/ReadingPreferencesSheet.swift
import SwiftUI

/// Bottom sheet for font, spacing, and TTS speed on content surfaces (M6.3).
struct ReadingPreferencesSheet: View {
    @Environment(AuthSession.self) private var session
    @Environment(\.colorScheme) private var colorScheme
    @Environment(\.dismiss) private var dismiss
    @Environment(\.readingPreferencesStore) private var store

    private let fontOptions: [(id: String, label: String)] = [
        ("default", "Default"),
        ("open-dyslexic", "OpenDyslexic"),
        ("atkinson", "Atkinson Hyperlegible"),
        ("system", "System font"),
    ]

    private let spacingOptions: [(id: String, label: String)] = [
        ("normal", "Normal"),
        ("wide", "Wide"),
        ("wider", "Wider"),
    ]

    var body: some View {
        NavigationStack {
            Form {
                Section("Font") {
                    Picker("Font face", selection: fontFaceBinding) {
                        ForEach(fontOptions, id: \.id) { option in
                            Text(option.label).tag(option.id)
                        }
                    }
                    .pickerStyle(.inline)
                    .labelsHidden()
                }

                Section("Letter spacing") {
                    Picker("Letter spacing", selection: letterSpacingBinding) {
                        ForEach(spacingOptions, id: \.id) { option in
                            Text(option.label).tag(option.id)
                        }
                    }
                    .pickerStyle(.segmented)
                    .labelsHidden()
                }

                Section("Word spacing") {
                    Picker("Word spacing", selection: wordSpacingBinding) {
                        ForEach(spacingOptions, id: \.id) { option in
                            Text(option.label).tag(option.id)
                        }
                    }
                    .pickerStyle(.segmented)
                    .labelsHidden()
                }

                Section("Line height") {
                    Picker("Line height", selection: lineHeightBinding) {
                        ForEach(spacingOptions, id: \.id) { option in
                            Text(option.label).tag(option.id)
                        }
                    }
                    .pickerStyle(.segmented)
                    .labelsHidden()
                }

                Section("Read aloud speed") {
                    Slider(value: ttsSpeedBinding, in: 0.5 ... 2.0, step: 0.1) {
                        Text("Speech rate")
                    }
                    Text(String(format: "%.1f×", store.prefs.ttsSpeed))
                        .font(.caption)
                        .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
                }
            }
            .navigationTitle("Reading preferences")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .confirmationAction) {
                    Button("Done") { dismiss() }
                }
            }
        }
        .presentationDetents([.medium, .large])
    }

    private var fontFaceBinding: Binding<String> {
        Binding(
            get: { store.prefs.fontFace },
            set: { newValue in
                Task {
                    await store.update(
                        ReadingPreferencesPatch(
                            fontFace: newValue,
                            dyslexiaDisplayEnabled: ReaderLogic.dyslexiaFromFontFace(newValue)
                        ),
                        accessToken: session.accessToken
                    )
                }
            }
        )
    }

    private var letterSpacingBinding: Binding<String> {
        Binding(
            get: { store.prefs.letterSpacing },
            set: { newValue in
                Task { await store.update(ReadingPreferencesPatch(letterSpacing: newValue), accessToken: session.accessToken) }
            }
        )
    }

    private var wordSpacingBinding: Binding<String> {
        Binding(
            get: { store.prefs.wordSpacing },
            set: { newValue in
                Task { await store.update(ReadingPreferencesPatch(wordSpacing: newValue), accessToken: session.accessToken) }
            }
        )
    }

    private var lineHeightBinding: Binding<String> {
        Binding(
            get: { store.prefs.lineHeight },
            set: { newValue in
                Task { await store.update(ReadingPreferencesPatch(lineHeight: newValue), accessToken: session.accessToken) }
            }
        )
    }

    private var ttsSpeedBinding: Binding<Double> {
        Binding(
            get: { store.prefs.ttsSpeed },
            set: { newValue in
                Task { await store.update(ReadingPreferencesPatch(ttsSpeed: newValue), accessToken: session.accessToken) }
            }
        )
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/features/reader/ReadingPreferencesSheet.kt
package com.lextures.android.features.reader

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.lextures.android.core.lms.ReadingPreferencesPatch

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ReadingPreferencesSheet(
    visible: Boolean,
    store: ReadingPreferencesStore,
    accessToken: String?,
    onDismiss: () -> Unit,
) {
    if (!visible) return
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val scope = rememberCoroutineScope()
    val fontOptions = listOf("default", "open-dyslexic", "atkinson", "system")
    val spacingOptions = listOf("normal", "wide", "wider")

    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = sheetState) {
        Column(
            modifier = Modifier.fillMaxWidth().padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Text("Reading preferences")
            Text("Font")
            fontOptions.forEach { face ->
                Button(
                    onClick = {
                        store.updateAsync(
                            scope,
                            ReadingPreferencesPatch(
                                fontFace = face,
                                dyslexiaDisplayEnabled = ReaderLogic.dyslexiaFromFontFace(face),
                            ),
                            accessToken,
                        )
                    },
                    modifier = Modifier.fillMaxWidth(),
                ) { Text(face.replaceFirstChar { it.uppercase() }) }
            }
            Text("Letter spacing")
            spacingOptions.forEach { spacing ->
                Button(
                    onClick = {
                        store.updateAsync(scope, ReadingPreferencesPatch(letterSpacing = spacing), accessToken)
                    },
                    modifier = Modifier.fillMaxWidth(),
                ) { Text(spacing) }
            }
            Text("Read aloud speed: ${"%.1f".format(store.row.ttsSpeed)}×")
            Slider(
                value = store.row.ttsSpeed.toFloat(),
                onValueChange = { speed ->
                    store.updateAsync(scope, ReadingPreferencesPatch(ttsSpeed = speed.toDouble()), accessToken)
                },
                valueRange = 0.5f..2f,
            )
            Button(onClick = onDismiss, modifier = Modifier.fillMaxWidth()) { Text("Done") }
        }
    }
}